### PR TITLE
Refactoring CLI to use config file, connect to Llama Index data sources, and allow for multiple agents 

### PR DESCRIPTION
**Pull request by sarahwooders**
Builds on top of #146 

Example running on hugging face dataset (loaded with Llama Index): 

image

# Updated CLI 

## Starting Agent 
```
> memgpt run [--agent <agent-name>] 
```

## Configuring defaults 
```
> memgpt configure 

? Do you want to enable MemGPT with Open AI? Yes
? Use OpenAI as default provider? Yes
? Do you want to enable MemGPT with Azure? No
? Select default model (recommended: gpt-4): gpt-4
['/Users/sarahwooders/repos/MemGPT/memgpt']
['sam_pov', 'memgpt_starter', 'memgpt_doc', 'sam_simple_pov_gpt35', 'sam', 'memgpt_doc', 'basic']
? Select default persona: sam_pov
['cs_phd', 'basic', 'basic']
? Select default human: cs_phd
Saving config to /Users/sarahwooders/.memgpt/config
```
## Loading External Data
External data connectors via Llama Index connectors implemented in `cli/cli_load.py`: 
```
memgpt load [directory/database] [ARGS]
```
This data source can now be attached to an agent by running `memgpt run --data-source <source-name>`.

## Listing Options (Agents, Humans, Personas, Data Sources) 

```
> memgpt list [agents/humans/personas/sources]

> memgpt list humans

+------------+---------------------------------------------------------------------------------------------+
|    Name    |                                             Text                                            |
+------------+---------------------------------------------------------------------------------------------+
| cs_phd.txt | This is what I know so far about the user, I should expand this as I learn more about them. |
|            |                                                                                             |
|            |                                       First name: Chad                                      |
|            |                                         Last name: ?                                        |
|            |                                         Gender: Male                                        |
|            |                                            Age: ?                                           |
|            |                                        Nationality: ?                                       |
|            |                   Occupation: Computer science PhD student at UC Berkeley                   |
|            |      Interests: Formula 1, Sailing, Taste of the Himalayas Restaurant in Berkeley, CSGO     |
| basic.txt  |                                       First name: Chad                                      |
|   basic    |                                       First name: Chad                                      |
+------------+---------------------------------------------------------------------------------------------+
```

## Adding Personas/Humans 
```
memgpt add [persona/human] --name <name> [-f <filepath>] [--text <text>]
```

## Directory Structure 
Configuration, credentials, agent state, and persona/human options are all saved in `~/.memgpt`. 
```
~/.memgpt
    /agents 
       /<agent_name> 
          config.json # agent config 
          /agent_state # message state
          /persistence_manager # persistence manager state
      ...
    /humans
        <human_name> # file with human persona 
        ...
    /personas
        <persona_name> # file with persona  
        ...
   /archival 
       /<index-name> # directory containing saved Llama Index 
        ...
   config # config file 
```

# Legacy Support 
Both legacy commands should still work: 
```
> memgpt 
> python main.py
```

# To-Dos 

- [x] Connect Llama Index embedding configuration with `MemGPTConfig` values
- [x] Graceful `Ctrl-D/Ctrl-C` exits 
- [x] Test agent checkpointing 
- [x] Beautify messages ✨

### CODE REVIEW

**Comment by cpacker** [comment]
If you say no to using OpenAI, and let the process crash out:

image

Then you run `memgpt run` afterwards and say "No" to select existing agent, it ends with:

image

instead of taking back to the first loop asking for an OpenAI key / if you want to use OpenAI

File: memgpt/main.py [hunk 1]
```diff
@@ -305,6 +326,7 @@ async def main(
 
     memgpt_agent = presets.use_preset(
         presets.DEFAULT,
```
Comment (cpacker) [inline memgpt/main.py:328]: @sarahwooders is there an easy way to allow settings "preset" via configure or a flag?

That way we can close #174

**Review by cpacker — CHANGES_REQUESTED** [review]

File: memgpt/cli/cli_config.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    default_provider = "openai"
+    MemGPTConfig.create_config_dir()
+
+    # openai credentials
+    use_openai = questionary.confirm("Do you want to enable MemGPT with Open AI?").ask()
+    if use_openai:
+        # search for key in enviornment
+        openai_key = os.getenv("OPENAI_API_KEY")
+        if not openai_key:
+            openai_key = questionary.text("Open AI keys not found in enviornment - please enter:").ask()
```
Comment (cpacker) [inline memgpt/cli/cli_config.py:32]: Typo should be "OpenAI API key"

File: memgpt/main.py (hunk 1 above)
Comment (sarahwooders) [reply memgpt/main.py:328]: Yeah we can add it as a line in both `AgentConfig` and `MemGPTConfig`, and set it up in `memgpt configure`. I can add a default option for now.

**Comment by sarahwooders** [comment]
@cpacker do you have a local model endpoint to test with? I fixed the hugging face embeddings model, so the vector index should work locally now.

File: memgpt/main.py (hunk 1 above)
Comment (sarahwooders) [reply memgpt/main.py]: Just pushed a change for this

**Pull request merged**